rfcs/ui-building-blocks.md

```diff
@@ -0,0 +1,413 @@
+      false => styles.remove(hover_style.entity),
+    }
+  }
+}
+```
+
+## Reference-level explanation
+
+### Style data flow
+
+At the heart of the styling design is a data flow that propagates style parameters from the style to the end widget.
+Naively, you'd like to just overwrite the parameter in question, applying the first style's value if any, then the next and so on.
+Unfortunately, this causes issues with dynamically applying and reverting styles, because the original value is lost completely.
+
+In order to get around this, we need to somehow duplicate our data, storing both the original and final values.
+This is done by creating two variants of each style parameter component: `Foo::Base` and `Foo::Final`.
+
+This requires the `StyleParam` trait:
+
+```rust
+trait StyleParam {
+  type Inner: Clone;
+
+  type Base: Component + DerefMut + Deref<Target = Self::Inner>;
+  type Final: Component + DerefMut + Deref<Target = Self::Inner>;
+}
+```
+
+End users should use the #[style_param] attribute macro on `struct Foo(...)`, which implements the `StyleParam` trait on an ordinary-looking struct.
+Under the hood, this creates the following boilerplate:
+
+```rust
+// PhantomData and the Inner type are used to ensure Foo is !Component
+// This avoids a footgun for users trying to query for Foo directly
+struct Foo(PhantomData<*mut u8>);
+
+// Actual data goes here
+struct FooInnerStyle(...)
+
+struct FooBase(FooInnerStyle);
+struct FooFinal(FooInnerStyle);
+
+/* add impls for deref/mut for both FooBase and FooFinal */
+
+impl StyleParam for Foo {
+  // Both Base and Final should derefence to the same type of data
+  // Allowing us to convert between them
+  type Inner = FooInnerStyle;
+
+  type Base = FooBase;
+  type Final = FooFinal;
+}
+```
+
+In order for the data to be propagated from our styles to our widgets, we need a set of **style propagation systems**, that work like so:
+
+```rust
+/// Rebuilds the style parameter `S` for the widget
+///
+/// Styles need to be update if either
+/// a) the styles associated with the widget has changed or b) the style's style parameter values have changed
+/// Styles are rebuilt from scratch, working from their base value each time the styles are changed
+fn apply_styles<S:StyleParam>(mut widget_param: &mut S, style_params: Vec<Option<&S>>){  
+  // If the style is set in that style, use style_param.set() to apply it
+  // This replaces any previous values for the `final` field
+}
+
+/// Automatically updates the styling for all style parameter components of type `S` whose Styles changed
+///
+/// End users should register this system in your app using `app.add_style::<S>()
+pub fn propagate_styles<S: StyleParam>(mut widget_query: Query<(&S::Base, &mut S::Final, &Styles), 
+  (With<Widget, Without<Style>, Changed<Styles>>)>,
+  style_query: Query<Option<&S::Base>, With<Style>>){
+ 
+ for (widget_param, styles) in widget_query.iter_mut(){
```

Comment:
`widget_query` is a `Query` of 3 things (`Base`, `Final`, and `Styles`), this iterator loop might be a bit of a typo.
It would have to be
```rust
for (sbase, sfinal, styles) in widget_query.iter_mut()
``` 
(also see below in `propagate_style_changes`)

Reply:
You're correct: this should be fixed up.

It's mostly pseudo-code, but we might as well fix it so it's clearer to readers :)